src/module-server/node-resolve.test.ts

```diff
@@ -0,0 +1,222 @@
+    const fs = await createFs(`
+      ./node_modules/preact/package.json {}
+      ./node_modules/preact/index.js
+      ./node_modules/preact/hooks/package.json {"main": "../dist/hooks/index.js"}
+      ./node_modules/preact/dist/hooks/index.js
+    `);
+    expect(await fs.resolve('preact')).toBe('./node_modules/preact/index.js');
+    expect(await fs.resolve('preact/hooks')).toBe(
+      './node_modules/preact/dist/hooks/index.js',
+    );
+  });
+});
+
+describe('resolving relative paths', () => {
+  test('implicitly adds .js extension', async () => {
+    const fs = await createFs(`
+      ./other.js
+    `);
+    expect(await fs.resolve('./other')).toBe('./other.js');
+  });
+
+  test('implicitly adds .ts extension', async () => {
+    const fs = await createFs(`
+      ./other.ts
+    `);
+    expect(await fs.resolve('./other')).toBe('./other.ts');
+  });
+
+  test('implicitly adds .tsx extension', async () => {
+    const fs = await createFs(`
+      ./other.tsx
+    `);
+    expect(await fs.resolve('./other')).toBe('./other.tsx');
+  });
+
+  test('does not implicitly add .mjs or .cjs', async () => {
+    const fs = await createFs(`
+      ./other.mjs
+      ./other.cjs
+    `);
+    expect(await fs.resolve('./other')).toBe(undefined);
+  });
+
+  test('extensions can be loaded explicitly', async () => {
+    const fs = await createFs(`
+      ./other.mjs
+      ./other.cjs
+      ./other.js
+      ./other.jsx
+      ./other.ts
+      ./other.tsx
+    `);
+    expect(await fs.resolve('./other.mjs')).toBe('./other.mjs');
+    expect(await fs.resolve('./other.cjs')).toBe('./other.cjs');
+    expect(await fs.resolve('./other.js')).toBe('./other.js');
+    expect(await fs.resolve('./other.jsx')).toBe('./other.jsx');
+    expect(await fs.resolve('./other.ts')).toBe('./other.ts');
+    expect(await fs.resolve('./other.tsx')).toBe('./other.tsx');
+  });
+
+  test('implicitly resolves folder/index.js', async () => {
+    const fs = await createFs(`
+      ./folder/index.js
+    `);
+    expect(await fs.resolve('./folder')).toBe('./folder/index.js');
+  });
+
+  test('implicitly resolves folder/index.ts', async () => {
+    const fs = await createFs(`
+      ./folder/index.ts
+    `);
+    expect(await fs.resolve('./folder')).toBe('./folder/index.ts');
+  });
+
+  test('does not implicitly resolve folder/index.cjs or folder/index.mjs', async () => {
+    const fs = await createFs(`
+      ./folder/index.mjs
+      ./folder/index.cjs
+    `);
+    expect(await fs.resolve('./other')).toBe(undefined);
+  });
+
+  test('resolves folder with package.json in it with main/module field, and ignores exports field', async () => {
+    const fs = await createFs(`
+      ./folder/package.json {"main": "./cjs.js", "module": "./mjs.js", "exports": "./dont-use.js"}
+      ./folder/mjs.js
+    `);
+    expect(await fs.resolve('./folder')).toBe('./folder/mjs.js');
+  });
+});
+
+test.todo('LOAD_PACKAGE_SELF');
+test.todo('LOAD_PACKAGE_IMPORTS');
```

Comment:
I left these tests out for now. Package imports are node's new way of creating "aliases" to files in your project, so you can do stuff like `import "#utils` instead of putting large relative paths. I didn't implement support for that yet. And package self refers to being able to import a package in itself (like how in this project we `import "pleasantest"` instead of `import '../dist/...'`). Support for that also isn't implemented yet.

Both are edge cases and I'll get to them eventually, or when someone asks for it